packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,436 @@
+  }
+
+  @override
+  void didChangeDependencies() {
+    _resolveImage();
+    super.didChangeDependencies();
+  }
+
+  @override
+  void didUpdateWidget(FadeInImage oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.image != oldWidget.image || widget.placeholder != widget.placeholder)
+      _resolveImage();
+  }
+
+  @override
+  void reassemble() {
+    _resolveImage(); // in case the image cache was flushed
+    super.reassemble();
+  }
+
+  void _resolveImage() {
+    _imageResolver.resolve(widget.image);
+    _placeholderResolver.resolve(widget.placeholder);
+
+    if (_phase == FadeInImagePhase.start)
+      _updatePhase();
+  }
+
+  void _updatePhase() {
+    setState(() {
+      switch(_phase) {
+        case FadeInImagePhase.start:
+          if (_imageResolver._imageInfo != null)
+            _phase = FadeInImagePhase.completed;
+          else
+            _phase = FadeInImagePhase.waiting;
+          break;
+        case FadeInImagePhase.waiting:
+          if (_imageResolver._imageInfo != null) {
+            // Received image data. Begin placeholder fade-out.
+            _controller.duration = widget.fadeOutDuration;
+            _animation = new CurvedAnimation(
+              parent: _controller,
+              curve: widget.fadeOutCurve,
+            );
+            _phase = FadeInImagePhase.fadeOut;
+            _controller.reverse(from: 1.0);
+          }
+          break;
+        case FadeInImagePhase.fadeOut:
+          if (_controller.status == AnimationStatus.dismissed) {
+            // Done fading out placeholder. Begin target image fade-in.
+            _controller.duration = widget.fadeInDuration;
+            _animation = new CurvedAnimation(
+              parent: _controller,
+              curve: widget.fadeInCurve,
+            );
+            _phase = FadeInImagePhase.fadeIn;
+            _controller.forward(from: 0.0);
+          }
+          break;
+        case FadeInImagePhase.fadeIn:
+          if (_controller.status == AnimationStatus.completed) {
+            // Done finding in new image.
+            _phase = FadeInImagePhase.completed;
+          }
+          break;
+        case FadeInImagePhase.completed:
+        // Nothing to do.
+          break;
+      }
+    });
+  }
+
+  @override
+  void dispose() {
+    _imageResolver.dispose();
+    _placeholderResolver.dispose();
+    _controller.dispose();
+    super.dispose();
+  }
+
+  ImageInfo get _imageInfo {
+    switch(_phase) {
+      case FadeInImagePhase.start:
+      case FadeInImagePhase.waiting:
+      case FadeInImagePhase.fadeOut:
+        return _placeholderResolver._imageInfo;
+      case FadeInImagePhase.fadeIn:
+      case FadeInImagePhase.completed:
+        return _imageResolver._imageInfo;
+    }
+
+    throw new StateError('Unrecognized FadeInImage phase: $_phase');
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    assert(_phase != FadeInImagePhase.start);
+    return new RawImage(
```

Comment:
It is cached in the `_ImageProviderResolver`. It only changes once when we switch from the placeholder to the target image.